specs/plasma.md

```diff
@@ -0,0 +1,199 @@
+# Plasma mode
+
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+**Table of Contents**
+
+- [Overview](#overview)
+- [DA Storage](#da-storage)
+- [Input Commitment](#input-commitment)
+- [Data Availability Challenge Contract](#data-availability-challenge-contract)
+  - [Parameters](#parameters)
+- [Derivation](#derivation)
+- [Safety and Finality](#safety-and-finality)
+- [Security Considerations](#security-considerations)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Overview
+
+Using [Plasma][vitalikblog] mode means switching to an offchain data availability provider.
+The protocol guarantees availability with a challenge contract on L1 where users
+can challenge the availability of input data used to derive the chain.
+Challenging an input commitment means forcing providers to submit the input data on chain
+to prove the data is available during a time period. We call the period during which any party
+can challenge a commitment `challenge_window` and the period during which any party
+can submit the input data to resolve the challenge `resolve_window`.
+If a challenge is not resolved by the time `resolve_window` of L1 blocks have passed,
+chain derivation must be reset, omitting the input data when rederiving therefore causing a reorg.
+
+## DA Storage
+
+Input data is uploaded to the storage layer via plain HTTP calls to the DA storage service.
+This service is horizontally scalable and concerned with replicating the data across prefered storage layers
+such as IPFS or any S3 compatible storage. Input data is content addressed by its hash in the request url
+so responses are easily cachable.
+
+Any DA provider can implement the following endpoints to receive and serve input data:
+
+- ```text
+  Request:
+    POST /put/<hex_encoded_commitment>
+    Content-Type: application/octet-stream
+    Body: <preimage_bytes>
+  
+  Response:
+    200 OK
+  ```
+
+- ```text
+  Request:
+    GET /get/<hex_encoded_commitment>
+  
+  Response:
+    200 OK
+    Content-Type: application/octet-stream
+    Body: <preimage_bytes>
+  ```
+
+## Input Commitment
+
+The batching and compression of input data remain unchanged. When a batch is ready
+to be submitted to the inbox address, the data is uploaded to the DA storage layer instead, and a
+commitment (keccak256 hash) is submitted as the bacher inbox transaction call data.
+The batcher SHOULD NOT submit a commitment onchain unless input data was successfully stored on the service.
```

Comment:
DA layer storage is not a consistent thing, you could get a `200` response just as the DA suffers some outage. In general, this outcome is going to be unavoidable. The outcome will be that a challenger will be able to win a reorg against the commitment. This is fine, as it's an expected part of the challenge game, DA Providers are responsible for the inputs. But I think it would be good to call out the outcome here, seeing as a violation of this `SHOULD NOT` has a potentially large impact.

Could just be as simple as saying "commitments made on chain without DA available inputs are subject to challenges. Read the \<challenge section\> for the impact of this"